pkg/controller/elasticache/user/setup.go

```
@@ -0,0 +1,200 @@
+			break
+		}
+	}
+	return resp
+}
+
+func (e *hooks) preCreate(ctx context.Context, cr *svcapitypes.User, obj *svcsdk.CreateUserInput) error {
+	obj.UserId = awsclients.String(meta.GetExternalName(cr))
+
+	if !awsclients.BoolValue(cr.Spec.ForProvider.NoPasswordRequired) {
+		pwds, err := elasticache.GetPasswords(ctx, e.kube, cr.Spec.ForProvider.PasswordSecretRef)
+		if resource.IgnoreNotFound(err) != nil {
+			return errors.Wrap(err, "cannot get password from the given secret")
+		}
+		obj.Passwords = pwds
+
+	}
+
+	return nil
+}
+
+func postCreate(_ context.Context, cr *svcapitypes.User, obj *svcsdk.CreateUserOutput, cre managed.ExternalCreation, err error) (managed.ExternalCreation, error) {
+	if err != nil {
+		return managed.ExternalCreation{}, err
+	}
+	meta.SetExternalName(cr, awsclients.StringValue(obj.UserId))
+	return cre, nil
+}
+
+func preObserve(_ context.Context, cr *svcapitypes.User, obj *svcsdk.DescribeUsersInput) error {
+	obj.UserId = awsclients.String(meta.GetExternalName(cr))
+	return nil
+}
+
+func postObserve(_ context.Context, cr *svcapitypes.User, obj *svcsdk.DescribeUsersOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+
+	switch awsclients.StringValue(obj.Users[0].Status) {
+	case "active":
+		cr.SetConditions(xpv1.Available())
+	case "modifying":
+		cr.SetConditions(xpv1.Unavailable())
+	case "deleting":
+		cr.SetConditions(xpv1.Deleting())
+	}
+
+	return obs, nil
+}
+
+func preDelete(_ context.Context, cr *svcapitypes.User, obj *svcsdk.DeleteUserInput) (bool, error) {
+	obj.UserId = awsclients.String(meta.GetExternalName(cr))
+	return false, nil
+}
+
+func postDelete(_ context.Context, cr *svcapitypes.User, obj *svcsdk.DeleteUserOutput, err error) error {
+	if err != nil {
+		if strings.Contains(err.Error(), svcsdk.ErrCodeInvalidUserStateFault) {
+			// skip: failed to delete User: InvalidUserState: User has status deleting.
```

Is there a reason why we silently drop this error?

If the user is already in a deleting state, it is better to check that during `preDelete` and skip the delete call in this case.

---

if we drop - the resource is in deletion - we drop the resource in postDelete - we don't need to wait .. its fine ?

---

I would prefer to check if the resource can be deleted in `preDelete` - if that is possible.